### DISCUSSION TITLE
Carelink delivery suspended state, and possibly other states

### DISCUSSION THREAD

**Discussion by kukuchta** [NightscoutFoundation/xDrip#3680] [Ideas]
From the point of view of the Carelink Follower user, there are some binary states in the downloaded data that could be added to the main chart. For my 740G pump, the one important is the Delivery Suspended state that marks the time periods when the insulin dosing is shut off due to the low sensor readings.

I imagine there are also some additional states for the hybrid 780g pump, like the temporary BG target, smart guard activation periods and so on. Im just guessing here. Not sure if there are such binary states for other data sources.

Im experimenting with this feature and wanted ask for opinion on how it is aligned with xDrip codebase and general concept.

First I added another SQL table and model called Notifications. I felt it was necessary, because currently Carelink notifications are stored in Treatment table as a treatment entry with a note but without any carbs or insulin dose. This seems more like an workaround than actual way of handling pump alerts and notifications. Every notification would have its own Type and boolean value.
Such Notifications are quite generic and do not need to be specific for Carelink. They would be parsed and populated by any data source.

Then, those Notifications would be used in BgGraphBuilder to conditionally produce some additional Lines. Each Notification with given notification Type produces two datapoints for given Line. If binary state is turned on, we take the timestamp and we add points with y=0, then y=highMark (simulating the rising edge). If the state is turned of, we add those points in different order (falling edge). This approach is also quite generic and would be usable with other data sources.

Do you think such changes are useful? Or there is too little gain for given amount of changes?

For Delivery Stopped indication my result looks like this (additional orange regions):
![obraz](https://github.com/user-attachments/assets/9fd1a3d1-5aab-40ce-8bde-b829f496991b)

**Comment by user-hash** [comment]
It would be very useful to have basal insulin displayed in xDrip using Carelink follower.
I wonder if we can display basal insulin delivery like on AAPS? 

![{38099F88-B661-441D-A027-6FA93C352687}](https://github.com/user-attachments/assets/d2d3b556-0749-4093-b133-05672285450c)

I have checked the Carelink report and it displays basal insulin in a similar way, I am just not sure how the collector works or how to extract the values.
Do you have any Carelink collector errors in your xDrip log?

**Comment by mortenfyhn** [comment]
I was about to start a new discussion but this is very similar so I'll try here.

I've been working on getting Minimed pump state (delivery suspended, temporary target, smartguard on/off) displayed on my Pebble watch. What I have now is the pump state as a string stored as a new field in the PumpStatus class. Then I just send that string to the watch and display it verbatim. It's not very pretty but it works.

It'd be nice to make this general enough so it's easy to extend to any pump with whatever states they have. 

I think an EnumSet could work well. The string is certainly not the best when you add more states.

My first step here is just getting the current state visible on the watch, and PumpStatus is good for that. But for the graph stuff the data needs to be added to the Treatments class or something like that right? Even on the Pebble watch it would be cool and useful to display this data in the graph (I've tweaked the graph to fill the Pebble's screen, so I have a generous 144 B&W pixels to work with).

**Reply by mortenfyhn** [reply]
Using an EnumSet for this seems to work ok. I'll try it out for a while: https://github.com/mortenfyhn/xDrip/commit/0d4658ec4e1c3d125ca293875747ba4af04cd758

**Reply by mortenfyhn** [reply]
Have used this for a while now. EnumSet for PumpState stored inside PumpStatus. Grabbing data from Minimed notification / companion app. Works well. When I get everything cleaned up and corner cases handled properly I'll make a PR.

![P1380298](https://github.com/user-attachments/assets/4200b99a-968f-4fd9-8e3b-8c8ba93c75c2)
